rtengine/badpixels.cc

```diff
@@ -445,126 +446,227 @@ int RawImageSource::interpolateBadPixelsXtrans(const PixelsMap &bitmapBads)
 /*  Search for hot or dead pixels in the image and update the map
  *  For each pixel compare its value to the average of similar color surrounding
  *  (Taken from Emil Martinec idea)
- *  (Optimized by Ingo Weyrich 2013 and 2015)
+ *  (Optimized by Ingo Weyrich 2013, 2015 and 2019)
  */
 int RawImageSource::findHotDeadPixels(PixelsMap &bpMap, const float thresh, const bool findHotPixels, const bool findDeadPixels) const
 {
+    BENCHFUN
     const float varthresh = (20.0 * (thresh / 100.0) + 1.0) / 24.f;
 
-    // allocate temporary buffer
-    float* cfablur = new float[H * W];
-
     // counter for dead or hot pixels
     int counter = 0;
 
 #ifdef _OPENMP
-    #pragma omp parallel
+    #pragma omp parallel reduction(+:counter)
 #endif
     {
+        array2D<float> cfablur(W, 5);
+        // zero left and right border
+        for (int i = 0; i < 5; ++i) {
+            cfablur[i][0] = cfablur[i][1] = cfablur[i][W - 2] = cfablur[i][W - 1];
+        }
+        int firstRow = -1;
+        int lastRow = -1;
+
 #ifdef _OPENMP
-        #pragma omp for schedule(dynamic,16) nowait
+        // note, static scheduling is important in this implementation
+        #pragma omp for schedule(static) nowait
 #endif
 
         for (int i = 2; i < H - 2; i++) {
-            for (int j = 2; j < W - 2; j++) {
+            if (firstRow == -1) {
+                firstRow = i;
+                if (firstRow == 2) {
```

Comment:
The scheduling splits the processing in chunks. Let's assume 800 rows and 8 threads. Because the scheduling is static we get 8 chunks of (in this case) equal size, each having its own i and so its own firstrow (0, 100, 200, ... , 700)